sdk/go/dagger/client.go

```diff
@@ -3,41 +3,136 @@ package dagger
 
 import (
 	"context"
-	"errors"
-	"net"
-	"net/http"
+	"os"
 
 	"github.com/Khan/genqlient/graphql"
+	"github.com/vektah/gqlparser/v2/gqlerror"
+	"go.dagger.io/dagger/sdk/go/dagger/api"
+	"go.dagger.io/dagger/sdk/go/dagger/engineconn"
+	_ "go.dagger.io/dagger/sdk/go/dagger/engineconn/embedded" // embedded connection
+	_ "go.dagger.io/dagger/sdk/go/dagger/engineconn/unix"     // unix connection
 )
 
-type SecretID string
+// Client is the Dagger Engine Client
+type Client struct {
+	conn engineconn.EngineConn
+	gql  graphql.Client
+}
+
+// ClientOpt holds a client option
+type ClientOpt func(cfg *engineconn.Config)
```

Comment:
Yes, I remember your suggestion and was planning to do that.

However I got a bit stuck. I don't want to expose engineconn to the user and originally intended for this to be dagger.Config{}, however got into a circular dependency problem (dagger -> engineconn -> dagger).

Do you see an easy way out of this?

(ps: I wish to move engineconn and whatever else is not part of the public API under internal/ so we're free to change without breaking anyone)

Reply:
I guess the generic fallback option w/ circular deps is to split out into a separate package, like:
* dagger -> engineconfig
* engineconn -> engineconfig

engineconfig is just a placeholder name, but it would just hold the common (public) interface/struct definitions needed to configure the engine. `engineconn` can remain internal too then I think.

I'd be fine with that. Intermediate option is to just change `ClientOpt` to be an interface rather than direct func so that we can retain the flexibility to do something like this in the future.

Reply:
> I'd be fine with that. Intermediate option is to just change ClientOpt to be an interface rather than direct func so that we can retain the flexibility to do something like this in the future.

Done.

Might find a better way than a separate package at some point, just don't want to it in this PR for now